File: oriedita/src/main/java/com/example/oriedita/editor/handler/MouseHandlerDisplayLengthBetweenPoints2.java
package com.example.oriedita.editor.handler;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.inject.Named;
import oriedita.editor.canvas.CreasePattern_Worker;
import oriedita.editor.canvas.MouseMode;
import oriedita.editor.databinding.MeasuresModel;
import origami.crease_pattern.OritaCalc;
import origami.crease_pattern.elements.LineSegment;
import origami.crease_pattern.elements.Point;

@ApplicationScoped
@Handles(MouseMode.DISPLAY_LENGTH_BETWEEN_POINTS_2_54)
public class MouseHandlerDisplayLengthBetweenPoints2 extends BaseMouseHandlerInputRestricted {
    private final MeasuresModel measuresModel;

    @Inject
    public MouseHandlerDisplayLengthBetweenPoints2(@Named("mainCreasePattern_Worker") CreasePattern_Worker d, MeasuresModel measuresModel) {
        this.d = d;
        this.measuresModel = measuresModel;
    }

    //マウス操作(ボタンを押したとき)時の作業
    public void mousePressed(Point p0) {
        Point p = d.getCamera().TV2object(p0);
        Point closest_point = d.getClosestPoint(p);
        if (p.distance(closest_point) < d.getSelectionDistance()) {
            d.lineStepAdd(new LineSegment(closest_point, closest_point, d.getLineColor()));
        }
    }

    //マウス操作(ドラッグしたとき)を行う関数
    public void mouseDragged(Point p0) {
    }

    //マウス操作(ボタンを離したとき)を行う関数
    public void mouseReleased(Point p0) {
        if (d.getLineStep().size() == 2) {
            measuresModel.setMeasuredLength2(OritaCalc.distance(d.getLineStep().get(0).getA(), d.getLineStep().get(1).getA()) * (double) d.getGrid().getGridSize() / 400.0);
            d.getLineStep().clear();
        }
    }
}

File: oriedita-ui/src/main/java/com/example/oriedita_ui/handler/MouseHandlerDisplayLengthBetweenPoints2.kt
package com.example.oriedita_ui.handler

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.PointerInputChange
import com.example.oriedita_core.origami.crease_pattern.elements.LineSegment
import com.example.oriedita_core.origami.crease_pattern.elements.Point
import com.example.oriedita_core.origami.crease_pattern.OritaCalc

/**
 * Обработчик измерения расстояния между точками (тип 2)
 * Адаптированная версия для Android
 */
class MouseHandlerDisplayLengthBetweenPoints2 : BaseMouseHandler() {
    
    private var isProcessing = false
    private var measuredLength = 0.0
    private var selectedPoints = mutableListOf<Point>()
    
    override fun onPress(offset: Offset, change: PointerInputChange): Boolean {
        val point = offsetToPoint(offset)
        val closestPoint = getClosestPoint(point)
        
        if (point.distance(closestPoint) < getSelectionDistance()) {
            val lineSegment = LineSegment(closestPoint, closestPoint, getLineColor())
            addLineStep(lineSegment)
            selectedPoints.add(closestPoint)
            
            println("Выбрана точка ${selectedPoints.size}: (${closestPoint.x}, ${closestPoint.y})")
        }
        
        return true
    }
    
    override fun onDrag(offset: Offset, change: PointerInputChange): Boolean {
        // Этот обработчик не требует перетаскивания
        return false
    }
    
    override fun onRelease(offset: Offset, change: PointerInputChange): Boolean {
        if (getLineStepSize() == 2) {
            calculateLength()
            clearLineStep()
            selectedPoints.clear()
        }
        
        return true
    }
    
    private fun calculateLength() {
        if (getLineStepSize() >= 2) {
            val point1 = (getLineStep(0) as LineSegment).getA() // Первая точка
            val point2 = (getLineStep(1) as LineSegment).getA() // Вторая точка
            
            val rawDistance = OritaCalc.distance(point1, point2)
            val gridSize = (getGrid()?.getGridSize() ?: 1.0).toDouble()
            measuredLength = rawDistance * gridSize / 400.0

            println("Измерено расстояние (тип 2): ${String.format("%.4f", measuredLength)}")
            println("Между точками:")
            println("  Точка 1: (${point1.x}, ${point1.y})")
            println("  Точка 2: (${point2.x}, ${point2.y})")
            println("  Сырое расстояние: ${String.format("%.4f", rawDistance)}")
            println("  Размер сетки: ${gridSize}")
        }
    }
    
    fun getMeasuredLength(): Double = measuredLength
    
    fun getSelectedPoints(): List<Point> = selectedPoints.toList()
    
    fun getSelectedPointsCount(): Int = selectedPoints.size
    
    fun isProcessing(): Boolean = isProcessing
    
    fun clearSelectedPoints() {
        selectedPoints.clear()
    }
    
    fun getLengthDescription(): String {
        return when (getLineStepSize()) {
            0 -> "Выберите первую точку"
            1 -> "Выберите вторую точку"
            2 -> "Расстояние измерено: ${String.format("%.4f", measuredLength)}"
            else -> "Неизвестное состояние"
        }
    }
    
    fun getLengthInfo(): String {
        return buildString {
            val gridSize = (getGrid()?.getGridSize() ?: 1.0).toDouble()
            append("Выбранные точки: ${selectedPoints.size}\n")
            if (selectedPoints.isNotEmpty()) {
                selectedPoints.forEachIndexed { index, point ->
                    append("  Точка ${index + 1}: (${String.format("%.2f", point.x)}, ${String.format("%.2f", point.y)})\n")
                }
            }
            append("Измеренное расстояние (тип 2): ${String.format("%.4f", measuredLength)}\n")
            append("Размер сетки: $gridSize")
        }
    }
    
    override fun getName(): String = "Измерение расстояния между точками (тип 2)"
    
    override fun getDescription(): String = "Измеряйте расстояния между двумя выбранными точками (второй тип)"
} 